### PR TITLE
fix: [WIP] Run rule engine only for associated program stages [DHIS2-10822]

### PR DESCRIPTION
**Pull request by zubaira**

### CODE REVIEW

**Review by enricocolasante — APPROVED** [review]

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java [hunk 1]
```diff
@@ -77,7 +78,14 @@ public ImplementableRuleService( ProgramRuleService programRuleService )
 
         List<ProgramRule> programRulesByActionTypes = getProgramRulesByActionTypes( program, programStageUid );
 
-        getProgramRulesCache().put( program.getUid(), !programRulesByActionTypes.isEmpty() );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java:80]: I think this cache could have been called `programHasRulesCache` to make it more descriptive.

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java [hunk 2]
```diff
@@ -77,7 +78,14 @@ public ImplementableRuleService( ProgramRuleService programRuleService )
...
+        if ( programStageUid == null )
+        {
+            getProgramRulesCache().put( program.getUid(), !programRulesByActionTypes.isEmpty() );
+
+            return programRulesByActionTypes.stream().filter( rule -> rule.getProgramStage() == null )
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java:85]: Could you maybe add a description of the logic here? We first cache whether a program has program rules with any program stage, and then return only rules with no program stage?

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java (hunk 1 above)
Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java:80]: resolved

File: dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java (hunk 2 above)
Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-program-rule/src/main/java/org/hisp/dhis/programrule/engine/ImplementableRuleService.java:88]: Added.
At enrollment, only those rules should be sent for execution with are NOT associated with any ProgramStage.

**Review by gnespolino — APPROVED** [review]

**Pull request merged**